### PR TITLE
Truncate outbox using an explicit range of sequence ids

### PR DESCRIPTION
**Pull request by pcholakov**
Ensure we don't create gaps in the outbox messages even in cases when truncation hints may get missed. This change updates the Partition Processor state machine to explicitly track the lower bound of the outbox, and truncate it based on a range of keys rather than the latest processed sequence number.

The truncation command assumes that it is operating on a reasonably narrow range as it performs essentially a counted loop and deletes all the keys within the range `[previous_outbox_high_watermark .. latest_truncation_sequence_id]`. No effort is made to ensure that this is a reasonably-sized interval.

Closes: https://github.com/restatedev/restate/issues/1642

### CODE REVIEW

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [inline crates/partition-store/src/outbox_table/mod.rs:21]: I changed this without too much thought since it more accurately captures the intent behind the truncate operation but I only afterwards saw that unlike the regular `Range`, the `start` and `end` fields aren't public, which requires a method call. I'm guessing it was a deliberate choice to use `Range`?

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [inline crates/partition-store/src/outbox_table/mod.rs:110]: I'd rather not merge this approach at all but curious if someone has a better way to implement this if we accept that scan-and-delete is actually an acceptable strategy here.

File: crates/storage-api/src/outbox_table/mod.rs
Comment (pcholakov) [inline crates/storage-api/src/outbox_table/mod.rs:62]: Probably overthinking this but I wonder if `trim_outbox` isn't a better name for operations of this type. Truncate to my mind is more closely associated with completely emptying a data structure, whereas "trim" is often used with logs to drop unwanted prefix data leaving the most recent entries intact.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (tillrohrmann) [inline crates/partition-store/src/outbox_table/mod.rs:113]: The outbox could actually be a candidate for https://github.com/facebook/rocksdb/wiki/Single-Delete which would avoid the propagation of the tombstones to the lowest level.

Comment (tillrohrmann) [reply crates/partition-store/src/outbox_table/mod.rs:113]: This is unrelated to this PR, though.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this draft PR @pcholakov. Happy to see that you got everything running and added some more tests :-) I left a few minor comments/questions.

File: crates/storage-api/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/storage-api/src/outbox_table/mod.rs:62]: Ah! This is useful context, thanks for the pointer!

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs:110]: Not at all important nor intended. I adopted the scan-and-delete (anti)pattern because 1) I didn't find the `delete_range_cf` operation on `StorageAccess` (nee `rocksdb::DBAccess`), but also 2) we don't (yet) track of the lower bound of unpruned outbox sequence ids.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs:108]: My memory is that there is no difference but this not based on any recent experience and certainly no testing. This is just replacement for `delete_range_cf` which I couldn't find a way to access - I should have called that out!

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs:113]: TIL! Going to read up on this, thank you for the link!

File: crates/partition-store/src/outbox_table/mod.rs
Comment (tillrohrmann) [reply crates/partition-store/src/outbox_table/mod.rs:110]: Regarding 2): The API expects a range and we are limiting the scan to this range. So why not directly deleting every entry between this range. Whether the range is correctly set or not, is then an orthogonal problem.

Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs:110]: That's absolutely viable! I was being overly paranoid about making a counted loop over a large range initially but given this is likely always going to be on the order of a range of 1-2 sequence ids, it's absolutely unnecessary. Let me revise that - with this optimisation in place, the PR as a whole should be viable to merge.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs]: Updated to an explicit counted loop.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs]: Now tracking low/high sequence ids explicitly and truncating between the two.

File: crates/partition-store/tests/outbox_table_test/mod.rs
Comment (pcholakov) [inline crates/partition-store/tests/outbox_table_test/mod.rs:59]: I realized that we're not testing the scenario where we want to truncate a range, but keep some messages behind. There was an off-by-one error in test code since it was deleting up to sequence_id `4` rather than '3' - hence the explicit split between outbox message count and the max sequence id encountered.

File: crates/worker/src/partition/state_machine/command_interpreter/mod.rs
Comment (pcholakov) [inline crates/worker/src/partition/state_machine/command_interpreter/mod.rs:115]: Not super happy with this naming but I couldn't come up with anything better; I'd like to rename this and the tail to be explicit - will revisit before merging.

File: crates/worker/src/partition/state_machine/effects.rs
Comment (pcholakov) [inline crates/worker/src/partition/state_machine/effects.rs:383]: Should we log the lower bound too?

File: crates/partition-store/src/outbox_table/mod.rs
Comment (tillrohrmann) [inline crates/partition-store/src/outbox_table/mod.rs:94]: Out of curiosity: Why did you change the explicit for loop into a map-for_each?

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for updating this PR @pcholakov. The changes look really good. I think there is only a problem with the initial value of the `outbox_head_seq_number`. Maybe this is an indicator that we are lacking test coverage for the scenario where we start with a filled outbox.

File: crates/partition-store/tests/outbox_table_test/mod.rs
Comment (tillrohrmann) [reply crates/partition-store/tests/outbox_table_test/mod.rs:59]: I know that the `partition-store` tests are a bit rigid. But instead of making the existing tests more complicated by testing some aspects as part of some bigger tests, I would suggest creating a dedicated test for the case you want to cover (e.g. truncating a single message).

File: crates/worker/src/partition/state_machine/effect_interpreter.rs
Comment (pcholakov) [reply crates/worker/src/partition/state_machine/effect_interpreter.rs:302]: Please ignore! Personal note that slipped my self-review. Should not be committed.

```suggestion
```

File: crates/partition-store/tests/outbox_table_test/mod.rs
Comment (pcholakov) [reply crates/partition-store/tests/outbox_table_test/mod.rs:63]: I made a comment elsewhere - I want to add explicit test coverage for partial truncation of the outbox to ensure that messages that are not targeted by the truncation command remain.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs:94]: I wrote the loop from scratch for the second PR revision, but I see now that the original was much simpler! I've moved it back to the for-loop form with the next iteration. I am just enjoying having sane iterator APIs after spending too long in JS land ;-)

File: crates/worker/src/partition/state_machine/command_interpreter/mod.rs
Comment (pcholakov) [reply crates/worker/src/partition/state_machine/command_interpreter/mod.rs:201]: Fair! I wanted to provide some context here for why we convert a single index into a range, but you're right that it's not this module's business to know how things work upstream. Will fix.

File: crates/worker/src/partition/state_machine/command_interpreter/mod.rs
Comment (pcholakov) [reply crates/worker/src/partition/state_machine/command_interpreter/mod.rs:144]: My mistake! I completely misread this. I'll add some test coverage for starting up with some messages already in the outbox.

File: crates/worker/src/partition/state_machine/command_interpreter/mod.rs
Comment (pcholakov) [reply crates/worker/src/partition/state_machine/command_interpreter/mod.rs:213]: Yes, absolutely – that's a bug! I'll add test coverage for `TruncateOutbox` handling to catch this.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [inline crates/partition-store/src/outbox_table/mod.rs:66]: I also tried the following approach (copied from `get_next_outbox_message`):

```rs
storage.get_first_blocking(
    TableScan::KeyRangeInclusiveInSinglePartition(partition_id, start, end),
    |kv| {
        if let Some((k, _)) = kv {
            let t = decode_key_value(k, &[])?;
            Ok(Some(t.0))
        } else {
            Ok(None)
        }
    },
)
```

But it failed with:

```
--- STDERR:              restate-partition-store::integration_test test_read_write ---
!!!! Runnning with test-util enabled !!!!
thread 'test_read_write' panicked at crates/partition-store/tests/outbox_table_test/mod.rs:43:10:
should not fail: Conversion(failed reading codec: remaining bytes in buf '0' < version bytes '1'

Stack backtrace:
   0: std::backtrace_rs::backtrace::libunwind::trace
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1: std::backtrace_rs::backtrace::trace_unsynchronized
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2: std::backtrace::Backtrace::create
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/backtrace.rs:331:13
   3: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
             at /Users/pavel/.cargo/registry/src/index.crates.io-6f17d22bba15001f/anyhow-1.0.81/src/backtrace.rs:27:14
   4: <T as core::convert::Into<U>>::into
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/convert/mod.rs:759:9
   5: restate_partition_store::outbox_table::decode_value::{{closure}}
             at ./src/outbox_table/mod.rs:229:51
   6: core::result::Result<T,E>::map_err
             at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/result.rs:829:27
   7: restate_partition_store::outbox_table::decode_value
             at ./src/outbox_table/mod.rs:228:26
   8: restate_partition_store::outbox_table::decode_key_value
             at ./src/outbox_table/mod.rs:221:26
   9: restate_partition_store::outbox_table::get_outbox_head_seq_number::{{closure}}
             at ./src/outbox_table/mod.rs:73:25
  10: restate_partition_store::partition_store::StorageAccess::get_first_blocking
             at ./src/partition_store.rs:686:9
  11: restate_partition_store::outbox_table::get_outbox_head_seq_number
             at ./src/outbox_table/mod.rs:69:5
  12: restate_partition_store::outbox_table::<impl restate_storage_api::outbox_table::ReadOnlyOutboxTable for restate_partition_store::partition_store::RocksDBTransaction>::get_outbox_head_seq_number::{{closure}}
             at ./src/outbox_table/mod.rs:180:9
  13: integration_test::outbox_table_test::verify_outbox_head_seq_number::{{closure}}
             at ./tests/outbox_table_test/mod.rs:42:10
  14: integration_test::outbox_table_test::run_tests::{{closure}}
             at ./tests/outbox_table_test/mod.rs:102:54
```

I didn't spend too much time trying to work out why that is but curious if you have any insights into what's going on here?

File: crates/partition-store/tests/outbox_table_test/mod.rs
Comment (pcholakov) [inline crates/partition-store/tests/outbox_table_test/mod.rs:95]: These transactions look a bit clumsy; I tried to create a helper that takes a transaction body as a closure but it ends up being quite involved as the type isn't object-safe, requiring a bunch of extra scaffolding. It's not clear to me that we benefit from committing in the interim but it is more representative of real world use.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (pcholakov) [reply crates/partition-store/src/outbox_table/mod.rs]: I didn't think you're suggesting that we this in this PR - let me know if I misunderstood. I didn't see that we use this anywhere else and I don't fully grok the implications of switching TBH.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (tillrohrmann) [reply crates/partition-store/src/outbox_table/mod.rs:66]: Because you are passing `&[]` as the value. If you replace `decode_key_value` with `OutboxKey::deserialize_from(&mut Cursor::new(k))?` it should probably work.

File: crates/partition-store/src/outbox_table/mod.rs
Comment (tillrohrmann) [inline crates/partition-store/src/outbox_table/mod.rs:59]: The iterator which is returned from `iterator_from` should already be seeked.

File: crates/worker/src/partition/storage/mod.rs
Comment (tillrohrmann) [reply crates/worker/src/partition/storage/mod.rs:151]: It is not 100% consistent. For lookups most methods use the `load_` prefix while a few use `get_`. Given this, I think your method name seems ok. Will need to unify it at some later point.

File: crates/worker/src/partition/state_machine/command_interpreter/tests.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/command_interpreter/tests.rs:745]: Nice tests :-)

**Review by tillrohrmann — APPROVED** [review]
Thanks for fixing the outbox truncation @pcholakov. The changes look good to me. Well done :-) +1 for merging.

**Pull request merged**
